flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/optimize/program/FlinkDynamicPartitionPruningConverterProgram.java

```diff
@@ -0,0 +1,147 @@
+ * +- Exchange(distribution=[broadcast])
+ *    +- Calc(select=[xxx], where=[<(xxx, xxx)]) # Need have an arbitrary filter condition.
+ *       +- TableSourceScan(table=[[dim, filter=[]]], fields=[xxx, dim_key])
+ * }</pre>
+ */
+public class FlinkDynamicPartitionPruningConverterProgram
+        implements FlinkOptimizeProgram<BatchOptimizeContext> {
+
+    @Override
+    public RelNode optimize(RelNode root, BatchOptimizeContext context) {
+        DefaultRelShuttle shuttle =
+                new DefaultRelShuttle() {
+                    @Override
+                    public RelNode visit(RelNode rel) {
+                        if (!(rel instanceof Join)) {
+                            List<RelNode> newInputs = new ArrayList<>();
+                            for (RelNode input : rel.getInputs()) {
+                                RelNode newInput = input.accept(this);
+                                newInputs.add(newInput);
+                            }
+                            return rel.copy(rel.getTraitSet(), newInputs);
+                        }
+                        Join join = (Join) rel;
+                        if (!ShortcutUtils.unwrapContext(join)
+                                .getTableConfig()
+                                .get(
+                                        OptimizerConfigOptions
+                                                .TABLE_OPTIMIZER_DYNAMIC_FILTERING_ENABLED)) {
+                            return join;
+                        }
+
+                        JoinInfo joinInfo = join.analyzeCondition();
+                        RelNode leftSide = join.getLeft();
+                        RelNode rightSide = join.getRight();
+                        Join newJoin = join;
+                        boolean changed = false;
+                        if (DynamicPartitionPruningUtils.isDimSide(leftSide)) {
+                            if (join.getJoinType() != JoinRelType.RIGHT) {
+                                Tuple2<Boolean, RelNode> relTuple =
+                                        DynamicPartitionPruningUtils
+                                                .canConvertAndConvertDppFactSide(
+                                                        rightSide,
+                                                        joinInfo.rightKeys,
+                                                        leftSide,
+                                                        joinInfo.leftKeys,
+                                                        join);
+                                changed = relTuple.f0;
+                                newJoin =
+                                        join.copy(
+                                                join.getTraitSet(),
+                                                Arrays.asList(leftSide, relTuple.f1.accept(this)));
+                            }
+                        }
+                        if (DynamicPartitionPruningUtils.isDimSide(rightSide)) {
```

Comment:
if leftSide is dime side, do we need to check the right side agin ?

Reply:
> if leftSide is dime side, do we need to check the right side agin ?

Personally, this is unnecessary. I change it to an if-else branch